mm2src/mm2_main/src/lp_healthcheck.rs

```diff
@@ -0,0 +1,505 @@
+            | HealthcheckRpcError::MessageEncodingFailed { .. } => StatusCode::INTERNAL_SERVER_ERROR,
+        }
+    }
+}
+
+pub async fn peer_connection_healthcheck_rpc(
+    ctx: MmArc,
+    req: RequestPayload,
+) -> Result<bool, MmError<HealthcheckRpcError>> {
+    // When things go awry, we want records to clear themselves to keep the memory clean of unused data.
+    // This is unrelated to the timeout logic.
+    let address_record_exp = Duration::from_secs(ctx.health_checker.config.timeout_secs);
+
+    let target_peer_address = req.peer_address;
+
+    let p2p_ctx = P2PContext::fetch_from_mm_arc(&ctx);
+    if target_peer_address == p2p_ctx.peer_id().into() {
+        // That's us, so return true.
+        return Ok(true);
+    }
+
+    let message = HealthcheckMessage::generate_message(
+        &ctx,
+        target_peer_address,
+        false,
+        ctx.health_checker
+            .config
+            .message_expiration_secs
+            .try_into()
+            .map_err(|e: TryFromIntError| HealthcheckRpcError::Internal { reason: e.to_string() })?,
+    )
+    .map_err(|reason| HealthcheckRpcError::MessageGenerationFailed { reason })?;
+
+    let encoded_message = message
+        .encode()
+        .map_err(|e| HealthcheckRpcError::MessageEncodingFailed { reason: e.to_string() })?;
+
+    let (tx, rx): (Sender<()>, Receiver<()>) = oneshot::channel();
+
+    {
+        let mut book = ctx.health_checker.response_handler.lock().await;
+        book.clear_expired_entries();
+        book.insert(target_peer_address.to_string(), tx, address_record_exp);
+    }
+
+    broadcast_p2p_msg(
+        &ctx,
+        peer_healthcheck_topic(&target_peer_address),
+        encoded_message,
+        None,
+    );
+
+    let timeout_duration = Duration::from_secs(ctx.health_checker.config.timeout_secs);
+    Ok(rx.timeout(timeout_duration).await == Ok(Ok(())))
+}
+
+pub(crate) async fn process_p2p_healthcheck_message(ctx: &MmArc, message: mm2_libp2p::GossipsubMessage) {
+    macro_rules! try_or_return {
+        ($exp:expr, $msg: expr) => {
+            match $exp {
+                Ok(t) => t,
+                Err(e) => {
+                    log::error!("{}, error: {e:?}", $msg);
+                    return;
+                },
+            }
+        };
+    }
+
+    let data = try_or_return!(
+        HealthcheckMessage::decode(&message.data),
+        "Couldn't decode healthcheck message"
+    );
+
+    let sender_peer = data.sender_peer();
+
+    let ctx = ctx.clone();
+
+    // Pass the remaining work to another thread to free up this one as soon as possible,
+    // so KDF can handle a high amount of healthcheck messages more efficiently.
+    ctx.spawner().spawn(async move {
+        let my_peer_address = P2PContext::fetch_from_mm_arc(&ctx).peer_id().into();
+
+        if !data.is_received_message_valid(my_peer_address, &ctx.health_checker.config) {
+            log::error!("Received an invalid healthcheck message.");
```

Comment:
...which makes this error log a bit too excessive.

Seed nodes gonna get this messages all the time.

Reply:
They propagate the message not process it.

Reply:
That would be horrific if they were processing all the messages..